### DISCUSSION TITLE
Animate a Haskell Diagrams

### DISCUSSION THREAD

**Discussion by cdupont** [reanimate/reanimate#247] [Q&A]
Hello, 
I'm trying to create an animation out of a Haskell Diagram (following up on this issue: #241).
I tried this code (from @mpilgrem):
```
{-# LANGUAGE OverloadedStrings #-}
module Reanimate.Diagrams
  ( renderDiagram
  ) where

import Data.Text.Lazy (toStrict)
import Diagrams.Backend.SVG (SVG (SVG), Options (SVGOptions))
import Diagrams.Core.Types (Diagram)
import Diagrams.Core.Compile (renderDia)
import Diagrams.Size (absolute)
import Graphics.Svg.Core (renderText)
import Graphics.SvgTree (Tree, parseSvgFile)
import Reanimate.Svg.Unuse (unbox)

import Data.Maybe (maybe)

renderDiagram :: Diagram SVG -> Tree
renderDiagram d =
  let opts = SVGOptions absolute Nothing "" [] False
      e = renderDia SVG opts d
      t = toStrict $ renderText e
      err = error "Malformed SVG"
  in  maybe err unbox (parseSvgFile "" t)
```
and:
```
module Main where

import Diagrams.Core.Types (Diagram)
import Diagrams.Backend.SVG (B)
import Diagrams.TwoD.Ellipse (circle)

import Reanimate (addStatic, mkBackground, reanimate, staticFrame)
import Reanimate.Diagrams (renderDiagram)

main :: IO ()
main = reanimate $ addStatic (mkBackground "cyan")
                 $ staticFrame 1 $ renderDiagram dia

dia :: Diagram B
dia = circle 1 # lw 0.1 <> square 1 # lw 0.1 # (moveTo $ p2 (1, 1))
```

I get this:
![image](https://user-images.githubusercontent.com/484033/141676992-be95f12a-0586-4228-afee-986651c27736.png)

However, if I generate directly the diagram (without Reanimate):
```
main :: IO ()
main = mainWith $ bg white $ dia
```

I get this:
![image](https://user-images.githubusercontent.com/484033/141677096-9add71a2-22ef-4302-8a3a-004f2f3422c4.png)

So the result is better than before (see #247), but there is still a problem of line width...
The diagram also seems to be flipped vertically.
Thanks

**Comment by cdupont** [comment]
Any idea? @Lemmih @mpilgrem

**Reply by mpilgrem** [reply]
The need to flip vertically arises because SVG and `reanimate` adopt different conventions for the direction of the y axis. See: https://reanimate.readthedocs.io/en/latest/introduction/#the-canvas.

**Reply by mpilgrem** [reply]
On line width, I think the use of the `absolute` option gives you what you requested - a width of 0.1 (relative to a canvas that is 16 wide by 9 high. In `diagrams`, by default, a width of 0.1 means something else - see https://diagrams.github.io/doc/manual.html#measurement-units.

**Comment by cdupont** [comment]
OK, so I can just use `flipYAxis` to flip the vertical axis it seems.

Regarding lineWidth, what I understand is that reanimate has a fixed canvas of 16*9, right? So a width of 0.1 will fit more or less 1% of the screen height (0.1/9). 
While in Diagrams, the line width of, say, a square, is respective to the size of that square. So a square of size 10 with line width 0.1 will be much bigger.

Another thing, in the Diagrams picture above, the circle and square are "fitting" the whole image. While in Reanimate, they are small and of-center. Why?
It seems that in Diagrams, the image canvas is set to the graphic bounding box. Can I do that in Reanimate too?

**Reply by mpilgrem** [reply]
If we make use of `traceShowId` from `Debug.Trace` we can see the SVG that `diagrams` is producing:
~~~haskell
renderDiagram :: Diagram SVG -> Tree
renderDiagram d = 
  let opts = SVGOptions absolute Nothing "" [] False
      e = renderDia SVG opts d
      t = traceShowId $ toStrict $ renderText e  -- Add a trace
      err = error "Malformed SVG"
  in  maybe err unbox (parseSvgFile "" t)
~~~
and that SVG file looks like this (deleting the uninteresting parts of it):
~~~xml
<svg>
  <g stroke-width=\"0.1\">
    <path d=\"M 2.5000,1.0000 l -0.0000,-1.0000 h -1.0000 l -0.0000,1.0000 Z\" />
  </g>
  <g stroke-width=\"0.1\">
    <path d=\"M 2.0000,1.5000 c 0.0000,-0.5523 -0.4477,-1.0000 -1.0000 -1.0000c -0.5523,-0.0000 -1.0000,0.4477 -1.0000
      1.0000c -0.0000,0.5523 0.4477,1.0000 1.0000 1.0000c 0.5523,0.0000 1.0000,-0.4477 1.0000 -1.0000Z\" />
  </g>
</svg>
~~~
so, two paths, each of which starts with an absolute 'MoveTo' command `M` - for example `M 2.5, 1.0`. See https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d#path_commands.

The `reanimate` canvas (which is 16 wide and 9 high) has its origin at the centre - again, see https://reanimate.readthedocs.io/en/latest/introduction/#the-canvas. So, `M 2.5, 1.0` moves the pen a little to the right and a little above the centre of the canvas.

Can you transform the SVG so that it fills the `reanimate` canvas? Yes. You can use `Reanimate.Svg.BoundingBox.boundingBox :: Tree -> (Double, Double, Double, Double)` to yield the bound box and `Reanimate.Svg.Constructors.scale :: Double -> Tree -> Tree` and `Reanimate.Svg.Constructors.translate :: Double -> Double -> Tree -> Tree` to change its size and move to where it needs to be.

**Reply by mpilgrem** [reply]
For example (extract, and making use of `center` rather than `translate`):
~~~haskell
import Reanimate (SVG, addStatic, boundingBox, center, flipYAxis, mkBackground,
  reanimate, scale, staticFrame)
import Reanimate.Diagrams (renderDiagram)

main :: IO ()
main = reanimate $ addStatic (mkBackground "cyan")
                 $ staticFrame 1 dia'

dia :: Diagram B
dia = circle 1 # lw 0.1 <> square 1 # lw 0.1 # moveTo (p2 (1, 1))

dia' :: SVG
dia' = fillCanvas $ flipYAxis $ renderDiagram dia

fillCanvas :: SVG -> SVG
fillCanvas t = scale m $ center t
 where
  (_, _, w, h) = boundingBox t
  m = min (16/w) (9/h)
~~~

**Comment by cdupont** [comment]
Thank you so much @mpilgrem !
Your code works well.
However, the arrows did not rendered correctly:
![image](https://user-images.githubusercontent.com/484033/142880204-db87a51e-2f60-4189-8519-41f8b007e467.png)

I noticed that the "absolute" size spec cause this. By providing a non absolute size it works:
```
opts = SVGOptions (mkWidth 250) Nothing "" [] False
```

Total success:

![image](https://user-images.githubusercontent.com/484033/142881735-3cb6a238-0092-41d1-9598-2de213f107f0.png)


This is the functions I use:

```
main :: IO ()
main = do
   let dia :: Diagram D.B
       dia = circle 1 # lwG 0.01 # named "circle"
           <> square 1 # named "square" # lwG 0.01 # (moveTo $ p2 (1, 1))
       dia' = connect "circle" "square" dia
   reanimate $ addStatic (mkBackground "white") $ staticFrame 1 $ RA.renderDiagram dia'

renderDiagram :: Diagram SVG -> Tree
renderDiagram d =
  let opts = SVGOptions (mkWidth 250) Nothing "" [] False
      svg = renderDia SVG opts d
      textSvg = toStrict $ renderText svg
  in  case (parseSvgFile "" textSvg) of
        Just tree -> fillCanvas $ flipYAxis $ unbox tree
        Nothing -> error "Malformed SVG"

fillCanvas :: R.SVG -> R.SVG
fillCanvas t = scale m $ center t
 where
  (_, _, w, h) = boundingBox t
  m = min (16/w) (9/h)
```

Are you interested in a pull request?

**Reply by mpilgrem** [reply]
On pull requests, I am not a maintainer of the package, just a fellow user of it. I think you need to put any proposal to @Lemmih. The impression I have, though, is that `reanimate` has sought to avoid a dependency on the `diagrams-*` packages since module `Reanimate.Diagrams` was dropped in `reanimate-0.2.0.0`.

**Reply by mpilgrem** [reply]
As an aside, your stated intent is to use `diagrams` as your foundation. However, with simple diagrams, you might consider starting with `reanimate-svg` or `Reanimate.Svg.Constructors`. In respect of making arrows in SVG, this blog post may be of interest: http://thenewcode.com/1068/Making-Arrows-in-SVG. The constructor of `Graphics.SvgTree.Types.DrawAttributes` includes the markers referred to in that post.

**Comment by cdupont** [comment]
I think a new package should be added: reanimate-diagrams. This would avoid to pollute reanimate with dependencies.
Actually, I started with Reanimate.Svg.Constructors, but I soon got limited because I needed arrow heads, and also bézier curves. Since I knew Diagrams a little, I jumped the ship. 
My diagram is not complex, it's basically the one in this discussion: https://github.com/reanimate/reanimate/discussions/245. I want to animate it by making it appear little by little, following the arrows. I also want to rotate it in 3D space (arrows are 2D but the coordinates are 3D).
Do you think I should come back? Will I be limited regarding to animations?

**Reply by mpilgrem** [reply]
I am away from a laptop and cannot test my idea with code, but I think this would be much easier to make all with `reanimate` and `reanimate-svg`.

If you import an SVG via `diagrams`, you are going to have to decompose what you input to get to the individual `Tree` values you ultimately wish to be a function of time - and then recompose them.

It must be easier to build up the SVG you need from its components.

For each arrow, you'll be using a path and `partialSvg` to animate it. The arrow heads will be a (static) marker node in a definition node. All together, that is the SVG that will be a function of time. At least, that is the 2D part.

**Reply by mpilgrem** [reply]
Here is example code illustrating my idea above:

~~~haskell
module Main where

-- From package lens
import Control.Lens ((&), (.~), (?~))
-- From package reanimate-svg
import Graphics.SvgTree (ElementRef (Ref), HasDrawAttributes (markerEnd),
  LineJoin (JoinRound), Marker, MarkerOrientation (..), Number (Num),
  Origin (..), PathCommand (..), RPoint, Tree, WithDefaultSvg (defaultSvg),
  attrId, definitionTree, fillOpacity, groupChildren, markerElements,
  markerHeight, markerOrient, markerRefPoint, markerTree, markerWidth,
  strokeColor, strokeLineJoin, strokeWidth)
-- From package linear
import Linear (V2 (V2), (^/))
-- From package reanimate
import Reanimate (Animation, andThen, animate, mapA, mkColor, mkDefinitions,
  mkGroup, mkLine, mkPath, mkPathString, partialSvg, reanimate)
import Reanimate.Builtin.Documentation (docEnv)

main :: IO ()
main = reanimate $ withMarker arrowHead $ docEnv animation
 where
  p1 = V2 0 4
  p2 = V2 4 (-4)
  p3 = V2 (-4) (-4)
  a1 = wavyArrow 3 theta p1 p2
  a2 = wavyArrow 3 theta p2 p3
  a3 = wavyArrow 3 theta p3 p1
  theta = pi/6
  animation = a1 `andThen` a2 `andThen` a3

withMarker :: Marker -> Animation -> Animation
withMarker marker = mapA addMarker
 where
  addMarker svg = mkGroup [mkDefinitions [markerTree marker], svg]

wavyArrow :: Int -> Double -> RPoint -> RPoint -> Animation
wavyArrow n theta p1 p2 = arrow $ wavyPath n theta p1 p2

-- | Assumes the arrowhead marker will have id "arrowhead".
arrow :: Tree -> Animation
arrow tree = animate $ \t -> partialSvg t tree'
 where
  tree' = tree & markerEnd ?~ Ref "arrowhead"

-- | Yields a wavy path.
wavyPath :: Int -> Double -> RPoint -> RPoint -> Tree
wavyPath 0 _ (V2 x1 y1) (V2 x2 y2) = mkLine (x1, y1) (x2, y2)
wavyPath _ 0 (V2 x1 y1) (V2 x2 y2) = mkLine (x1, y1) (x2, y2)
wavyPath n theta p1 p2 = mkPath $ start <> middle <> end
 where
  start = [MoveTo OriginAbsolute [p1], CurveTo OriginRelative [(d1, d2, d3)]]
  middle = [SmoothCurveTo OriginRelative (replicate (n - 1) (d2, d3))]
  end = [SmoothCurveTo OriginRelative [(d3 - d1, d3)]]
  d3 = (p2 - p1) ^/ fromIntegral (n + 1)
  d1 = d3 ^/ 3
  d2 = d3 - rot d1
  rot (V2 x y) = V2 (x * ct - y * st) (x * st + y * ct)
  st = sin theta
  ct = cos theta

-- | Yields a marker with id "arrowhead"
arrowHead :: Marker
arrowHead = defaultSvg & markerHeight ?~ Num 4 & markerWidth ?~ Num 8 &
  markerRefPoint .~ (Num 6.5, Num 2) & markerOrient ?~ OrientationAuto &
  attrId ?~ "arrowhead" & fillOpacity ?~ 0 & strokeWidth ?~ Num 1 &
  strokeColor ?~ mkColor "black" & strokeLineJoin ?~ JoinRound &
  markerElements .~ [mkPathString "M 0.5 0.5 L 6.5 2 L 0.5 3.5"]
~~~

and the result:
![test](https://user-images.githubusercontent.com/14206448/143940221-a7d17067-ce3a-4409-b90b-e80e41ca620d.gif)

**Reply by cdupont** [reply]
Whouah that's excellent!

**Comment by cdupont** [comment]
Understood... I was hoping to pass a time parameter to the Diagram section, and build the diagram using (or up to) the time provided.
But maybe Diagram doesn't provide the necessary tools, such as `partialSvg`.
